### ISSUE TITLE
New lower-case name restriction may be incorrect for some uses

### ISSUE THREAD

**Issue by fasaxc** [projectcalico/libcalico-go#684]
We use the "name" validator in a lot of places in the v3 datamodel and the name now restricts to lower case:

```
v3$ ag -R 'validate.*name'
bgppeer.go
46:	Node string `json:"node,omitempty" validate:"omitempty,name"`

workloadendpoint.go
40:	Orchestrator string `json:"orchestrator,omitempty" validate:"omitempty,name"`
42:	Workload string `json:"workload,omitempty" validate:"omitempty,name"`
44:	Node string `json:"node,omitempty" validate:"omitempty,name"`
46:	ContainerID string `json:"containerID,omitempty" validate:"omitempty,name"`
48:	Pod string `json:"pod,omitempty" validate:"omitempty,name"`
50:	Endpoint string `json:"endpoint,omitempty" validate:"omitempty,name"`
68:	Profiles []string `json:"profiles,omitempty" validate:"omitempty,dive,name"`

hostendpoint.go
44:	Node string `json:"node,omitempty" validate:"omitempty,name"`
61:	Profiles []string `json:"profiles,omitempty" validate:"omitempty,dive,name"`
```

I think at least some of those fields may need a more loose validation:

- `containerID`: the CNI spec just says this will be "plain text"
- `Orchestrator` `Workload` `Pod`, `Endpoint`: not sure these are guaranteed lower-case
- `Profiles`: could be mixed case before v3

**Comment by caseydavenport** [comment]
> containerID: the CNI spec just says this will be "plain text"

Yes - I think this should be more permissive. 

> Orchestrator Workload Pod, Endpoint: not sure these are guaranteed lower-case

Yep. Pod is guaranteed to be lower-case, but the others may have mixed case, though for Workload and Orchestrator we have some control over the values that get put there.

> Profiles: could be mixed case before v3

If you mean the name, then all names for all resources are lowe-case only in Calico v3.

**Comment by bcreane** [comment]
Discussed with @caseydavenport and decided to prioritize mixed-case containerIDs in v3 validation. Goal is to avoid breaking k8s-friendly container orchestrations such as rkt that allow uppercase letters in containerID.

**Comment by caseydavenport** [comment]
To be clear, we haven't checked what values things like rkt and cri-o use for their container IDs.

If we can confirm they pass the validation then I don't think we need to do anything.

CC @gunjan5

**Comment by bcreane** [comment]
I did discuss with @gunjan5 prior to the earlier comment. He mentioned that rkt does in fact support mixed case container names. Quoth @gunjan5 

 "It doesn't make sense to be restrictive on something like containerID, especially in WEP spec
since the path containerID travels is orchestrator/kubelet -> CNI-plugin -> IPAM -> CNI -> libcalico/WEP/Datastore (where we validate)... which seems to be the last step, failing validation there feels wrong"

**Comment by gunjan5** [comment]
it's supposed to be a UUID. Seems like rkt uses this regex for their testing ``^[a-fA-F0-9-]{36}$` //UUID, close enough` (included the comment from their code)

**Comment by gunjan5** [comment]
cri-o doesn't seem to have anything other than being a `string` type

**Comment by caseydavenport** [comment]
ok, so we should probably relax our restriction on the ContainerID field for v3.0.

I think the other ones can wait until we find a compelling reason.

**Comment by caseydavenport** [comment]
We've addressed what we need to for v3.0.

I'll move to v3.1 to see if we need to do any of the others.

**Comment by caseydavenport** [comment]
I think the main thing remaining here is the continerID validation, which is tracked here: https://github.com/projectcalico/libcalico-go/issues/776
